addons/folo/common/src/main/java/org/commonjava/indy/folo/data/FoloRecordCassandra.java

```diff
@@ -0,0 +1,433 @@
+import org.commonjava.indy.folo.conf.FoloConfig;
+import org.commonjava.indy.folo.model.TrackedContent;
+import org.commonjava.indy.folo.model.TrackedContentEntry;
+import org.commonjava.indy.folo.model.TrackingKey;
+import org.commonjava.indy.subsys.cassandra.CassandraClient;
+import org.commonjava.indy.subsys.infinispan.CacheHandle;
+import org.commonjava.maven.galley.util.UrlUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.UriInfo;
+import java.net.MalformedURLException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@ApplicationScoped
+@FoloStoreToCassandra
+public class FoloRecordCassandra implements FoloRecord,StartupAction {
+
+    private final static String DOWNLOADS = "DOWNLOAD";
+    private final static String UPLOADS = "UPLOAD";
+
+    private final Logger logger = LoggerFactory.getLogger( this.getClass() );
+
+    @Inject
+    CassandraClient cassandraClient;
+
+    @Inject
+    FoloConfig config;
+
+    @Context
+    UriInfo uriInfo;
+
+    @FoloInprogressCache
+    @Inject
+    private CacheHandle<TrackedContentEntry, TrackedContentEntry> inProgressRecords;
+
+    @FoloSealedCache
+    @Inject
+    private CacheHandle<TrackingKey, TrackedContent> sealedRecords;
+
+    @Inject
+    private FoloBackupListener foloBackupListener;
+
+    @Inject
+    private FoloExpirationWarningListener expirationWarningListener;
+
+
+
+    private Session session;
+    private Mapper<DtxBuildRecord> buildsMapper;
+    private Mapper<DtxTrackingRecord> trackingMapper;
+
+    private PreparedStatement getTrackingRecordByBuildIdAndPath;
+    private PreparedStatement getTrackingRecordBySealed;
+    private PreparedStatement getTrackingRecordsByTrackingKey;
+
+
+    private static String createFoloBuildsTable( String keyspace )
+    {
+        return "CREATE TABLE IF NOT EXISTS " + keyspace + ".builds ("
+                + "id text,"
+                + "sealed boolean,"
+                + "started timestamp,"
+                + "finished timestamp,"
+                + "PRIMARY KEY (id)"
+                + ");";
+    }
+
+    private static String createFoloRecordsTable( String keyspace )
+    {
+        return "CREATE TABLE IF NOT EXISTS " + keyspace + ".records ("
+                + "tracking_key text,"
+                + "sealed boolean,"
+                + "store_key text,"
+                + "access_channel text,"
+                + "path text,"
+                + "origin_url text,"
+                + "local_url text,"
+                + "store_effect text,"
+                + "md5 text,"
+                + "sha256 text,"
+                + "sha1 text,"
+                + "size bigint,"
+                + "timestamps set<bigint>,"
+                + "PRIMARY KEY ((tracking_key),path)"
```

Comment:
I guess store_key and store_effect (download/upload) should be in the primary keys. ie., PRIMARY KEY ((tracking_key),store_key,store_effect,path)

Reply:
Doesn't need to because almost every request for folo records is based on tracking key.
Tracking_key as. partition key and path is enough to. provide uniqueness for. stored records based on build ids.

Reply:
We should probably have `storekey` and possibly `store_effect` as cluster keys (in addition to `path`) though, right? So we can filter more efficiently on those when we query, without having to resort to `ALLOW FILTERING`?

If we collapse the sealed records table into the path records table, then I guess there would be an index on `sealed` in that table?

Reply:
> "We should probably have storekey and possibly store_effect as cluster keys (in addition to path) though, right? So we can filter more efficiently on those when we query, without having to resort to ALLOW FILTERING?"

- We don't need them , tracking_key and path is enough. to provide us with uniqueness for every. record and if we need to query especially for them then we have option of creating index for every column that we need but that we don't need because every method request is based on TrackingKey.

> If we collapse the sealed records table into the path records table, then I guess there would be an index on sealed in that table?

- We have index on 'sealed'	reecords (indexing in cassandra. is. creating another hiden table) so. queries will be performant and we. don't have. to. use "ALLOW FILTERING" .
from cassandra. documentation about indexes:
""An index provides a means to access data in Cassandra using attributes other than the partition key. The benefit is fast, efficient lookup of data matching a given condition. The index indexes column values in a separate, hidden table from the one that contains the values being indexed.""